File: src/lib/qron-tokenomics.ts
/**
 * $QRON obligation math. Not settlement.
 *
 * Live chain: 1_000_000_000 supply, almost all held by the payTo EOA.
 * The 100_000_000 figure on govchain.us is theater. Do not use it.
 *
 * Two different percents must not be added:
 * - Supply split 40/30/15/15 is a proposed cap table, not the holder map.
 * - Fiat burn is 25% of an enterprise-revenue dollar.
 *   The other four buckets split only the remaining 75%, in that same
 *   40:30:15:15 ratio, so the parts sum to the dollar. Applying 40/30/15/15
 *   to the gross and then also taking 25% burn would book 125%.
 *
 * Nothing here buys, burns, or transfers tokens. Autonomy stays off.
 */

export const LIVE_QRON_SUPPLY = 1_000_000_000;
export const THEATER_SUPPLY_DO_NOT_USE = 100_000_000;

export const SUPPLY_SPLIT = {
  treasury: 0.4,
  nodeOperators: 0.3,
  coreContributors: 0.15,
  ecosystemGrants: 0.15,
} as const;

export const FIAT_BURN_RATE = 0.25;

export type RevenueObligations = {
  amountCents: number;
  burnCents: number;
  treasuryCents: number;
  nodeOperatorsCents: number;
  coreContributorsCents: number;
  ecosystemGrantsCents: number;
  settlesOnChain: false;
  autonomyEnabled: false;
};

export function computeRevenueObligations(
  amountFiatUsd: number
): RevenueObligations {
  if (!Number.isFinite(amountFiatUsd) || amountFiatUsd < 0) {
    throw new Error("amountFiatUsd must be a non-negative finite number");
  }
  const cents = Math.round(amountFiatUsd * 100);
  const burnCents = Math.round(cents * FIAT_BURN_RATE);
  const rest = cents - burnCents;
  const treasuryCents = Math.round(rest * SUPPLY_SPLIT.treasury);
  const nodeOperatorsCents = Math.round(rest * SUPPLY_SPLIT.nodeOperators);
  const coreContributorsCents = Math.round(
    rest * SUPPLY_SPLIT.coreContributors
  );
  const ecosystemGrantsCents =
    rest - treasuryCents - nodeOperatorsCents - coreContributorsCents;
  const sum =
    burnCents +
    treasuryCents +
    nodeOperatorsCents +
    coreContributorsCents +
    ecosystemGrantsCents;
  if (sum !== cents) {
    throw new Error(`obligation sum ${sum} != ${cents}`);
  }
  return {
    amountCents: cents,
    burnCents,
    treasuryCents,
    nodeOperatorsCents,
    coreContributorsCents,
    ecosystemGrantsCents,
    settlesOnChain: false,
    autonomyEnabled: false,
  };
}

File: agentz/core/qron_tokenomics.py
"""Simulated $QRON fiat-obligation math.

Mirrors src/lib/qron-tokenomics.ts. Does not trade, burn, or transfer.
Live supply is 1_000_000_000. The 100_000_000 govchain figure is theater.
Burn is 25% of the fiat dollar. 40/30/15/15 splits only the remaining 75%.
"""

from __future__ import annotations

LIVE_QRON_SUPPLY = 1_000_000_000
THEATER_SUPPLY_DO_NOT_USE = 100_000_000
FIAT_BURN_RATE = 0.25
SUPPLY_SPLIT = {
    "treasury": 0.40,
    "node_operators": 0.30,
    "core_contributors": 0.15,
    "ecosystem_grants": 0.15,
}


def compute_revenue_obligations(amount_fiat_usd: float) -> dict:
    if amount_fiat_usd < 0:
        raise ValueError("amount_fiat_usd must be non-negative")
    cents = int(round(amount_fiat_usd * 100))
    burn = int(round(cents * FIAT_BURN_RATE))
    rest = cents - burn
    treasury = int(round(rest * SUPPLY_SPLIT["treasury"]))
    nodes = int(round(rest * SUPPLY_SPLIT["node_operators"]))
    core = int(round(rest * SUPPLY_SPLIT["core_contributors"]))
    grants = rest - treasury - nodes - core
    total = burn + treasury + nodes + core + grants
    if total != cents:
        raise RuntimeError(f"obligation sum {total} != {cents}")
    return {
        "amount_cents": cents,
        "burn_cents": burn,
        "treasury_cents": treasury,
        "node_operators_cents": nodes,
        "core_contributors_cents": core,
        "ecosystem_grants_cents": grants,
        "settles_on_chain": False,
        "autonomy_enabled": False,
        "live_supply": LIVE_QRON_SUPPLY,
    }
